File: 2019/python/day09.py
from utils import read_data
from utils.intcode import VM


def p1(program):
    vm = VM(program[:], inputs=iter((1,)))
    outputs = list(vm.outputs())
    assert len(outputs) == 1
    return outputs[0]


def p2(program):
    vm = VM(program[:], inputs=iter((2,)))
    outputs = list(vm.outputs())
    assert len(outputs) == 1
    return outputs[0]


def run_tests():
    program1 = [
        int(x)
        for x in "109,1,204,-1,1001,100,1,100,1008,100,16,101,1006,101,0,99".split(",")
    ]
    assert all(a == b for a, b in zip(program1, VM(program1).outputs()))
    program2 = [int(x) for x in "1102,34915192,34915192,7,4,7,99,0".split(",")]
    assert len(str(next(VM(program2).outputs()))) == 16
    program3 = [int(x) for x in "104,1125899906842624,99".split(",")]
    assert next(VM(program3).outputs()) == 1125899906842624


if __name__ == "__main__":
    run_tests()
    print("All tests passed")

    data = read_data()
    program = [int(x) for x in data.strip().split(',')]
    print("Part 1: {}".format(p1(program)))
    print("Part 2: {}".format(p2(program)))

File: 2019/rust/src/bin/day09.rs
use std::iter;

use anyhow::{anyhow, bail, Result};

use aoc::intcode::{RegisterValue, VM};

fn main() -> Result<()> {
    env_logger::init();

    let input = aoc::utils::get_input()?;
    let program = aoc::intcode::parse_program(&input)?;

    println!("Part 1: {}", part1(&program)?);
    println!("Part 2: {}", part2(&program)?);
    Ok(())
}

fn part1(program: &[RegisterValue]) -> Result<RegisterValue> {
    run_with_single_input(program, 1)
}

fn part2(program: &[RegisterValue]) -> Result<RegisterValue> {
    run_with_single_input(program, 2)
}

fn run_with_single_input(program: &[RegisterValue], input: RegisterValue) -> Result<RegisterValue> {
    let mut vm = VM::from_iterator(program.to_vec(), iter::once(input));
    let value = vm.outputs().next().ok_or(anyhow!("No output generated"))?;
    if vm.outputs().next().is_some() {
        bail!("Unexpected output");
    }

    Ok(value)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn init() {
        let _ = env_logger::builder().is_test(true).try_init();
    }

    #[test]
    fn test_program_1() {
        init();

        let program = vec![
            109, 1, 204, -1, 1001, 100, 1, 100, 1008, 100, 16, 101, 1006, 101, 0, 99,
        ];
        let outputs: Vec<RegisterValue> = VM::from_iterator(program.clone(), std::iter::empty())
            .outputs()
            .collect();
        assert_eq!(outputs, program);
    }

    #[test]
    fn test_program_2() {
        init();

        let program = vec![1102, 34915192, 34915192, 7, 4, 7, 99, 0];
        let outputs: Vec<RegisterValue> = VM::from_iterator(program, std::iter::empty())
            .outputs()
            .collect();
        assert_eq!(outputs.len(), 1);
        assert!(outputs[0] >= 1_000_000_000_000_000 && outputs[0] <= 9_999_999_999_999_999);
    }

    #[test]
    fn test_program_3() {
        init();

        let program = vec![104, 112589906842624, 99];
        let outputs: Vec<RegisterValue> = VM::from_iterator(program.clone(), std::iter::empty())
            .outputs()
            .collect();
        assert_eq!(outputs[..], program[1..2]);
    }
}
